onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+        reduced_size *= left_dim;
+      } else if (has_left_dim) {  // if it is only in one of left and right, we can reduce right away
+        left = ReduceSum<T>(left, i, allocator, tp);
+      } else if (has_right_dim) {
+        right = ReduceSum<T>(right, i, allocator, tp);
+      }
+    } else {  // This dimension is not reduced (i.e.) it appears in the output after processing these 2 operands
+      // Both the left and right operands have non-trivial dimension value along this axis
+      // They must be equal
+      if (has_left_dim && has_right_dim) {
+        ORT_ENFORCE(left_dim == right_dim, "Einsum op: Input shapes do not align");
+        lro.push_back(i);
+        lro_size *= left_dim;
+      } else if (has_left_dim) {
+        // The left operand has non-trivial dimension value
+        lo.push_back(i);
+        lo_size *= left_dim;
+      } else {
+        // The right operand may or may not have non-trivial dim value
+        // If it has trivial dim value (1),
+        // it will just form a trailing dimension for the right operand
+        ro.push_back(i);
+        ro_size *= right_dim;
+      }
+    }
+  }
+
+  // Permutate the left operand so that the axes order go like this: [lro, lo, reduce_dims, ro]
+  std::vector<size_t> left_permutation;
+  left_permutation.reserve(lro.size() + lo.size() + reduce_dims.size() + ro.size());
+  left_permutation.insert(left_permutation.end(), lro.begin(), lro.end());
+  left_permutation.insert(left_permutation.end(), lo.begin(), lo.end());
+  left_permutation.insert(left_permutation.end(), reduce_dims.begin(), reduce_dims.end());
+  left_permutation.insert(left_permutation.end(), ro.begin(), ro.end());
+  if (IsTransposeRequired(left.Shape().GetDims().size(), left_permutation)) {
+    left = Transpose(left, left_permutation, allocator);
+  }
+  CreateReshapedView(left, {lro_size, lo_size, reduced_size});
+
+  // Permutate the right operand so that the axes order go like this: [lro, reduce_dims, ro, lo]
+  std::vector<size_t> right_permutation;
+  right_permutation.reserve(lro.size() + lo.size() + reduce_dims.size() + ro.size());
+  right_permutation.insert(right_permutation.end(), lro.begin(), lro.end());
+  right_permutation.insert(right_permutation.end(), reduce_dims.begin(), reduce_dims.end());
+  right_permutation.insert(right_permutation.end(), ro.begin(), ro.end());
+  right_permutation.insert(right_permutation.end(), lo.begin(), lo.end());
+  if (IsTransposeRequired(right.Shape().GetDims().size(), right_permutation)) {
+    right = Transpose(right, right_permutation, allocator);
+  }
+  CreateReshapedView(right, {lro_size, reduced_size, ro_size});
+
+  // Calculate output size
+  // Output shape will be, by rules of matmul:
+  // because we are multiplying two tensors of shapes [lro, lo, reduce_dims] , [lro, reduce_dims, ro]
+  // [dim_value of `lro` dims,
+  //  dim_value of `lo` dims,
+  // `1` for each of the `reduce_dims`,
+  // dim_value of `ro` dims]
```

Comment:
this is just to maintain the consistency of the rank until you get to the final pair right?

Reply:
yes we maintain ranks until the final PairwiseOperandProcess() call after which we only care about the final output and don't get it to the "homogenized" rank